pkg/csi/controller.go

```diff
@@ -2,53 +2,148 @@ package csi
+	}
+	s.config.driver.config.Logger.Info("Volume created",
+		zap.Strings("attrs", []string{volume.Attributes["repo"], volume.Attributes["hash"]}))
+	return &csi.CreateVolumeResponse{
+		Volume: &volume,
+	}, nil
 }
 
 func (s *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
-	return nil, nil
+	return &csi.DeleteVolumeResponse{}, nil
 }
 
 func (s *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
-	return nil, nil
+	return &csi.ValidateVolumeCapabilitiesResponse{
+		Supported: true,
+	}, nil
 }
 
 func (s *controllerServer) ControllerGetCapabilities(ctx context.Context, req *csi.ControllerGetCapabilitiesRequest) (*csi.ControllerGetCapabilitiesResponse, error) {
-	return nil, nil
+	s.config.driver.config.Logger.Sugar().Debugf("ControllerGetCapabilities: %+v", s.config.driver.controllerServiceCapabilities)
+	return &csi.ControllerGetCapabilitiesResponse{
+		Capabilities: s.config.driver.controllerServiceCapabilities,
+	}, nil
 }
 
 func (s *controllerServer) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
-	return nil, nil
+	return nil, status.Error(codes.Unimplemented, "ControllerPublishVolume unsupported")
 }
 
 func (s *controllerServer) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
-	return nil, nil
+	return nil, status.Error(codes.Unimplemented, "ControllerUnpublishVolume unsupported")
 }
 
 func (s *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
-	return nil, nil
+	s.config.driver.config.Logger.Debug("ListVolumes request", zap.Int32("max", req.MaxEntries), zap.String("token", req.StartingToken))
+	// skip to the token passed in.
+	repos, err := core.ListReposPaginated(s.config.driver.metadataStore, req.StartingToken)
+	if err != nil {
+		s.config.driver.config.Logger.Error("ListVolumes: failed to list repos", zap.Error(err))
+		return nil, fmt.Errorf("datamon failed to list repos, err:%v", err)
+	}
+	volumes, nextToken := getVolumeEntries(repos, req.MaxEntries)
+	// return a list of repos as volumes.
+	s.config.driver.config.Logger.Debug("ListVolumes request done", zap.Int("count", len(volumes)), zap.String("nextToken", nextToken))
+	return &csi.ListVolumesResponse{
+		Entries:   volumes,
+		NextToken: nextToken,
+	}, nil
+}
+
+func getVolumeEntries(repos []model.RepoDescriptor, max int32) ([]*csi.ListVolumesResponse_Entry, string) {
+	volumes := make([]*csi.ListVolumesResponse_Entry, max)
+	for _, repo := range repos {
+		if max == 0 {
+			return volumes, repo.Name
+		}
+		volume := csi.ListVolumesResponse_Entry{
+			Volume: &csi.Volume{
+				CapacityBytes: math.MaxInt64,
+				Id:            repo.Name,
```

Comment:
remain somewhat confused as to the difference between this `csi.Volume` and the use in `CreateVolume`.  in one function, CSI volumes to correspond to bundles (uniquely identified by repo name + root hash), whereas here, volumes correspond to repos.

Reply:
The datamon model is made to fit into this. In some ways, the volumes are repos and exact instance of the volume (snapshot) is the hash. This could have been modeled as snapshots but snapshot APIs are still in the works and needs each snapshot to be created. We might revisit this modeling when CSI is more baked and after getting feedback from the k8s community.

Reply:
> revisit this modeling when CSI is more baked and after getting feedback from the k8s community

it's a plan.  #160 is open to track `ListVolumesResponse` as it pertains to datamon.